Install Python and configure Debian Backports only if necessary

jsf9k opened the pull request:
## 🗣 Description ##

This pull request make the following changes:
- Moves the python, pip, and backports role dependencies from `meta/main.yml` to `meta/requirements.yml`
- Only installs the Python and Pip Ansible roles if actually necessary
- Only configures Debian Backports if actually necessary

## 💭 Motivation and context ##

There is no reason to install Python or Pip or configure Debian Backports if they are not being used.

## 🧪 Testing ##

All automated tests pass.  See also the testing done in cisagov/kali-packer#151.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Revert to using the default branch of [cisagov/ansible-role-python](https://github.com/cisagov/ansible-role-python) once cisagov/ansible-role-python#55 is merged.

jsf9k committed:
Move role dependencies to meta/requirements.yml

We will install pip, python, and backports only if needed.

jsf9k committed:
Only install python and pip when they are actually required

jsf9k committed:
Only configure for backports when it is actually required

jsf9k committed:
Use a non-default branch of cisagov/ansible-role-python for testing

This change can be reverted once cisagov/ansible-role-python#55 is
merged.

jsf9k committed:
Add a role variable for installing Python development dependencies

This is useful, e.g., if pip must build a wheel.

jsf9k committed:
Install Python development dependencies for ImpDump

This is required because in this case pip must build wheels from
source.

dav3r commented on tasks/install_python.yml:
```diff
@@ -0,0 +1,14 @@
+      ansible.builtin.include_role:
+        name: python
+      vars:
+        python_install_python2: "{{ assessment_tool_python2 }}"
+        python_install_development_dependencies: "{{ assessment_tool_python_install_development_dependencies }}"
```
Alphabetize:
```suggestion
        python_install_development_dependencies: "{{ assessment_tool_python_install_development_dependencies }}"
        python_install_python2: "{{ assessment_tool_python2 }}"
```

dav3r reviewed (approved):
👍 
![just-one-more-thing](https://github.com/cisagov/ansible-role-assessment-tool/assets/21343441/8a0e1eee-cfae-4820-8cdd-b99b2552b19b)

jsf9k committed:
Alphabetize task attributes

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj commented on defaults/main.yml:
This is just for visibility but I would really like to move away from `yes`/`no` and just stick to `true`/`false` when using Boolean values.

jsf9k replied:
Are you requiring me to make those changes here?  I would prefer to create a separate PR for that, which I can do as soon as this one is merged.

mcdonnnj replied:
That was just for visibility so an issue or separate PR is fine. I more wanted to raise it for consideration by the team in case there was opposition.

jsf9k replied:
Please see #49, which I will address once this PR is merged.

mcdonnnj commented on tasks/main.yml:
```diff
@@ -48,15 +48,15 @@
-- name: Create a Python virtualenv if necessary
-  ansible.builtin.include_tasks: create_python_venv.yml
+- name: Install Python and create Python virtualenv
   when:
     - assessment_tool_pip_packages is defined or assessment_tool_pip_requirements_file is defined
```
This logic now gates whether or not Python (both 2 and 3) is installed at all. Does this match the behavior we expect?

jsf9k replied:
The intent is that if the user does not specify a list of `pip` packages to install or a requirements file, then we don't create a virtual environment and hence have no reason to install Python.

Would you prefer that we (choose all that apply, and feel free to add more options if necessary):
- [x] Also install Python if `assessment_tool_python_install_development_dependencies` or `assessment_tool_python2` is defined and set to `true`?
- [x] Also install Python if a _new_ role variable, say, `assessment_tool_python` is defined and set to `true`?

jsf9k replied:
I went ahead and created cisagov/kali-packer#151 to identify any edge cases that we may have missed.

mcdonnnj replied:
I suppose the latter makes the most sense to me? My thinking is that a project with a Python script that has no external package dependencies would no longer be guaranteed to work with this change in place. I'm not sure how much it technically matters but I see it as breaking behavior.

jsf9k replied:
Please see commit b7305d4.  I ended up implementing _both_ choices because that makes the most sense to me.  If you specify _any_ Python role vars then Python will be installed, and if you don't need to specify any then you still have `assessment_role_python` to force it to happen.

mcdonnnj reviewed (commented):
One comment for visibility and then a question. Also now is a good time to revert the branch for [cisagov/ansible-role-python](https://github.com/cisagov/ansible-role-python).

jsf9k committed:
Revert "Use a non-default branch of cisagov/ansible-role-python for testing"

This reverts commit 2d4d926a16b0a288ce25ec846dbc4a1de3306003.

We can do this now that cisagov/ansible-role-python#55 has been
merged.

jsf9k commented:
> Also now is a good time to revert the branch for [cisagov/ansible-role-python](https://github.com/cisagov/ansible-role-python).

Please see commit e6d9536.

jsf9k committed:
Add a role variable that can force the installation of Python

This role variable also forces the creation of an empty Python virtual
environment.  This is occassionally useful, e.g., when installing a
tool that lacks any Python packaging and has no dependencies.

We also add a test case to the "Python" Molecule scenario that tests
the use of this new variable.

dav3r reviewed (approved):
Looks great 👍 👍

mcdonnnj commented on molecule/python/converge.yml:
```diff
@@ -50,3 +50,16 @@
         assessment_tool_install_dir: /tools/mitm6
         assessment_tool_pip_packages:
           - mitm6
+    - name: Install the sqlmap package in a venv
```
Since it's not organized as a Python package and we do not create a venv.

```suggestion
    - name: Install the sqlmap tool
```

mcdonnnj commented on molecule/python/tests/test_python.py:
```diff
@@ -18,6 +18,7 @@
         "/tools/sshenum",
         "/tools/dirsearch",
         "/tools/mitm6",
+        "/tools/sqlmap",
```
If it doesn't hurt anything to alphabetize.

```suggestion
        "/tools/dirsearch",
        "/tools/mitm6",
        "/tools/sqlmap",
        "/tools/sshenum",
```

mcdonnnj commented on molecule/python/tests/test_python.py:
```diff
@@ -49,6 +50,8 @@ def test_packages(host, pkg):
         ),
         ("/tools/mitm6/.venv", ["mitm6"]),
+        # There is no venv for sqlmap since it has no dependencies.
+        # ("/tools/sqlmap/.venv", []),
```
I'm not sure if you also wanted to alphabetize these parameters by the directory name.

jsf9k replied:
Please see commit b9ad57e.

mcdonnnj reviewed (approved):
This has shaped up nicely. Thanks for your work 🏆. Just some minor suggestions for your consideration.

jsf9k committed:
Edit task comment to make it more correct

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Alphabetize directories by name

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Alphabetize parameter sets by directory name

Also alphabetize pip package names where they are present.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
